compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Context.scala

```diff
@@ -83,60 +89,50 @@ class Context(val contextId: Context.ContextId) {
     dop.decodeScenarioModule(homePackageId, lfScenarioModule)
   }
 
-  private def validate(pkgIds: Traversable[PackageId]): Unit =
-    pkgIds.foreach(
-      Validation.checkPackage(allPackages, _).left.foreach(e => throw ParseError(e.pretty)),
-    )
-
   @throws[ParseError]
   def update(
-      unloadModules: Seq[String],
+      unloadModules: Set[ModuleName],
       loadModules: Seq[ProtoScenarioModule],
-      unloadPackages: Seq[String],
+      unloadPackages: Set[PackageId],
       loadPackages: Seq[ByteString],
       omitValidation: Boolean,
-  ): Unit = this.synchronized {
+  ): Unit = synchronized {
 
-    // First we unload modules and packages
-    unloadModules.foreach { moduleId =>
-      val lfModuleId = assert(ModuleName.fromString(moduleId))
-      modules -= lfModuleId
-      defns = defns.filterKeys(ref => ref.packageId != homePackageId || ref.modName != lfModuleId)
-    }
-    unloadPackages.foreach { pkgId =>
-      val lfPkgId = assert(PackageId.fromString(pkgId))
-      extPackages -= lfPkgId
-      defns = defns.filterKeys(ref => ref.packageId != lfPkgId)
-    }
-    // Now we can load the new packages.
     val newPackages =
       loadPackages.map { archive =>
         Decode.decodeArchiveFromInputStream(archive.newInput)
       }.toMap
-    extPackages ++= newPackages
-    defns ++= Compiler(extPackages).compilePackages(extPackages.keys, !omitValidation)
 
-    // And now the new modules can be loaded.
+    if (unloadPackages.nonEmpty || newPackages.nonEmpty) {
+      // if any change we recompile everything
+      extPackages --= unloadPackages
+      extPackages ++= newPackages
+      extDefns = Compiler(extPackages).compilePackages(extPackages.keys)
+    }
+
     val lfModules = loadModules.map(module =>
       decodeModule(LanguageVersion.Major.V1, module.getMinor, module.getDamlLf1))
 
-    modules ++= lfModules.map(m => m.name -> m)
+    modules --= unloadModules
+    lfModules.foreach(mod => modules += mod.name -> mod)
 
-    // At this point 'allPackages' is consistent and we can
-    // compile the new modules.
     val compiler = Compiler(allPackages)
-    defns = lfModules.foldLeft(defns)(
-      (newDefns, m) =>
-        newDefns.filterKeys(ref => ref.packageId != homePackageId || ref.modName != m.name)
-          ++ m.definitions.flatMap {
-            case (defName, defn) =>
-              compiler
-                .compileDefn(Identifier(homePackageId, QualifiedName(m.name, defName)), defn)
-        })
+    modDefns = HashMap.empty
+    modules.foreach {
+      case (modName, mod) =>
+        mod.definitions.foreach {
+          case (defName, defn) =>
+            modDefns ++= compiler
+              .compileDefn(Identifier(homePackageId, QualifiedName(modName, defName)), defn)
+        }
+    }
+
+    defns = modDefns orElse extDefns
```

Comment:
Afaik `orElse` can end up doing two map lookups whereas before we only did one. On the other hand if the first one succeeds, it is probably cheaper since the map is small. I guess the different performance characteristic here just doesn’t matter?

Reply:
In fact, the performance characteristic matter here. Thanks to underlying it. 
Any function call in the engine will be slower with this double lookup. 
I rewrite to remove this double lookup.